decoder/src/logger/mod.rs

```diff
@@ -63,98 +55,141 @@ pub fn log_defmt(
     );
 }
 
-struct Logger {
-    /// Whether to log `debug!` and `trace!`-level host messages.
-    verbose: bool,
+/// Determines whether `metadata` belongs to a log record produced by [`log_defmt`].
+pub fn is_defmt_frame(metadata: &Metadata) -> bool {
+    metadata.target().starts_with(DEFMT_TARGET_MARKER)
+}
 
-    /// Number of characters used by the timestamp. This may increase over time and is used to align
-    /// messages.
-    timing_align: AtomicUsize,
+/// A `log` record representing a defmt log frame.
+pub struct DefmtRecord<'a> {
+    timestamp: &'a str,
+    level: Level,
+    args: fmt::Arguments<'a>,
+    module_path: Option<&'a str>,
+    file: Option<&'a str>,
+    line: Option<u32>,
 }
 
-impl Log for Logger {
-    fn enabled(&self, metadata: &Metadata) -> bool {
-        if metadata.target().starts_with(DEFMT_TARGET_MARKER) {
-            // defmt is configured at firmware-level, we will print all of it.
-            true
-        } else {
-            // Host logs use `info!` as the default level, but with the `verbose` flag set we log at
-            // `trace!` level instead.
-            if self.verbose {
-                metadata.target().starts_with("probe_run")
-            } else {
-                metadata.target().starts_with("probe_run") && metadata.level() <= Level::Info
-            }
+impl<'a> DefmtRecord<'a> {
+    /// If `record` was produced by [`log_defmt`], returns the corresponding `DefmtRecord`.
+    pub fn new(record: &Record<'a>) -> Option<Self> {
+        let target = record.metadata().target();
+        let is_defmt = target.starts_with(DEFMT_TARGET_MARKER);
+        if !is_defmt {
+            return None;
         }
+
+        let timestamp = &target[DEFMT_TARGET_MARKER.len()..];
+
+        Some(Self {
+            level: record.level(),
+            timestamp,
+            args: *record.args(),
+            module_path: record.module_path(),
+            file: record.file(),
+            line: record.line(),
+        })
+    }
+
+    /// Returns the formatted defmt timestamp.
+    pub fn timestamp(&self) -> &str {
+        self.timestamp
+    }
+
+    pub fn level(&self) -> Level {
+        self.level
+    }
+
+    pub fn args(&self) -> &fmt::Arguments<'a> {
+        &self.args
+    }
+
+    pub fn module_path(&self) -> Option<&'a str> {
+        self.module_path
+    }
+
+    pub fn file(&self) -> Option<&'a str> {
+        self.file
     }
 
-    fn log(&self, record: &Record) {
-        if !self.enabled(record.metadata()) {
-            return;
+    pub fn line(&self) -> Option<u32> {
+        self.line
+    }
+
+    /// Returns a builder that can format this record for displaying it to the user.
+    pub fn print(&'a self) -> Printer<'a> {
```

Comment:
nit: `print()` sounds like it would print itself, maybe `printer()` would make it clearer that this is not the case?